pkg/proxy/nftables/proxier.go

```diff
@@ -1639,20 +1655,31 @@ func (proxier *Proxier) writeServiceToEndpointRules(tx *knftables.Transaction, s
 			ipX = "ip6"
 		}
 
-		for _, ep := range endpoints {
-			epInfo, ok := ep.(*endpointInfo)
-			if !ok {
-				continue
-			}
+		// Rule to map an known source IP to an endpoint IP.
+		tx.Add(&knftables.Rule{
+			Chain: svcChain,
+			Rule: knftables.Concat(
+				ipX, "daddr set", ipX, "saddr map", "@", svcInfo.affinityMapToEPName,
+			),
+		})
+
+		// Rule to look up endpoint chain from an endpoint IP.
+		tx.Add(&knftables.Rule{
+			Chain: svcChain,
+			Rule: knftables.Concat(
+				ipX, "daddr vmap", "@", svcInfo.affinityMapToChainName,
+			),
+		})
+
+		// Fallback rule to delete affinity IP associated with an unavailable endpoint IP
+		// due to endpoint scale-in or failure
+		tx.Add(&knftables.Rule{
+			Chain: svcChain,
+			Rule: knftables.Concat(
+				ipX, "daddr !=", svcInfo.ClusterIP(), "delete", "@", svcInfo.affinityMapToEPName, "{", ipX, "saddr : ", ipX, "daddr }",
```

Comment:
svcChain can be shared by LB IP, External IP, and NodePort. Does the reason why it needs to exclude ClusterIP apply to other Service IPs?

Reply:
Good catching. The check should be "NOT cluster-cidr" in this fallback rule